src/Fabulous/AttributeDefinitions.fs

```diff
@@ -3,135 +3,215 @@
 open System.Collections.Generic
 open Fabulous
 
-type IAttributeDefinition =
-    abstract member Key: AttributeKey
-    abstract member UpdateNode: obj voption -> obj voption -> IViewNode -> unit
-
-type IScalarAttributeDefinition =
-    inherit IAttributeDefinition
-    abstract member CompareBoxed: a: obj -> b: obj -> ScalarAttributeComparison
+[<Struct>]
+type ScalarAttributeData =
+    { UpdateNode: obj voption -> obj voption -> IViewNode -> unit
+      CompareBoxed: obj -> obj -> ScalarAttributeComparison }
+
+[<Struct>]
+type SmallScalarAttributeData =
+    { UpdateNode: uint64 voption -> uint64 voption -> IViewNode -> unit }
+
+[<Struct>]
+type WidgetAttributeData =
+    { ApplyDiff: WidgetDiff -> IViewNode -> unit
+      UpdateNode: Widget voption -> Widget voption -> IViewNode -> unit }
 
+[<Struct>]
+type WidgetCollectionAttributeData =
+    { ApplyDiff: ArraySlice<Widget> -> WidgetCollectionItemChanges -> IViewNode -> unit
+      UpdateNode: ArraySlice<Widget> voption -> ArraySlice<Widget> voption -> IViewNode -> unit }
 
 /// Attribute definition for scalar properties
+[<Struct>]
 type ScalarAttributeDefinition<'inputType, 'modelType, 'valueType> =
-    { Key: AttributeKey
+    { Key: ScalarAttributeKey
       Name: string
-      Convert: 'inputType -> 'modelType
-      ConvertValue: 'modelType -> 'valueType
-      Compare: 'modelType -> 'modelType -> ScalarAttributeComparison
-      UpdateNode: 'valueType voption -> 'valueType voption -> IViewNode -> unit }
+      Convert: 'inputType -> 'modelType }
 
-    member x.WithValue(value) : ScalarAttribute =
+    member inline x.WithValue(value: 'inputType) : ScalarAttribute =
         { Key = x.Key
 #if DEBUG
           DebugName = x.Name
 #endif
+          NumericValue = 0UL
           Value = x.Convert(value) }
```

Comment:
Yes, about 5% for the benchmark I have, I suspect the difference might be even more pronounced on mobile devices because they have less CPU cache. It is a dynamic dispatch call that is on the hot path (attribute creation)

Reply:
about `AttributeDefinition` and `AttributeData`

`AttributeDefinition` is for creating attributes instances. E.g. `WithValue` method, in essence it is our public API

`AttributeData` is for accessing **untyped** functions for diffing and applying diffs, it is used only in the reconciler.  Note that all of them are `[<Structs>]`, thus it removed one more pointer chasing when accessing it.

I'm not sure what are the right names for them are but one is a part of public API and the other is used internally in the engine.

Reply:
Also `AttributeData` is stored in separate respective arrays for faster access (accessing a function is a single memory jump from the array pointer)